File: MetaGeta/MGDataStore/FileSet.vb
<Serializable()> _
Public Class Bucket
    Inherits FileSet

    Private m_Tag As MGTag

    Public Sub New(ByVal tag As MGTag)
        m_Tag = tag
    End Sub

    Public ReadOnly Property Tag() As MGTag
        Get
            Return m_Tag
        End Get
    End Property

End Class

<Serializable()> _
Public Class FileSet
    Implements IEnumerable(Of MGFile)

    Protected m_Items As List(Of MGFile)

    Public Sub New()
        m_Items = New List(Of MGFile)()
    End Sub

    Public Sub New(ByVal c As IEnumerable(Of MGFile))
        m_Items = New List(Of MGFile)(c)
        m_Items.Sort(New MGFileIDComparer())
    End Sub

    Public Sub Add(ByVal file As MGFile)
        m_Items.Add(file)
        m_Items.Sort(New MGFileIDComparer())
    End Sub

    Public Function GetAllTags(ByVal tagName As String) As IEnumerable(Of MGTag)
        Dim tags As New List(Of MGTag)
        For Each file As MGFile In m_Items
            tags.Add(file.Tags.Item(tagName))
        Next
        Return From t In tags Group By t.Value Into tg = First() Select tg
    End Function

    Public Function GetEnumerator() As System.Collections.Generic.IEnumerator(Of MGFile) Implements IEnumerable(Of MGFile).GetEnumerator
        Return m_Items.GetEnumerator()
    End Function
    Public Function GetEnumerator1() As System.Collections.IEnumerator Implements IEnumerable.GetEnumerator
        Return m_Items.GetEnumerator()
    End Function

    Public Shared Function Intesect(ByVal left As FileSet, ByVal right As FileSet) As FileSet
        Dim common As New FileSet

        Dim l As Integer = 0, r As Integer = 0
        While (l < left.m_Items.Count AndAlso r < right.m_Items.Count)
            If left.m_Items(l).ID = right.m_Items(r).ID Then
                common.m_Items.Add(left.m_Items(l))
                l += 1
                l += 1
            ElseIf left.m_Items(l).ID.CompareTo(right.m_Items(r).ID) < 0 Then
                l += 1
            Else
                r += 1
            End If
        End While

        Return common
    End Function
End Class

File: MetaGeta/MetaGeta.DataStore/Model/FileSet.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections;
using System.Collections.Generic;

#endregion

namespace MetaGeta.DataStore {
    [Serializable]
    public class FileSet : IEnumerable<MGFile> {
        protected List<MGFile> m_Items;

        public FileSet() {
            m_Items = new List<MGFile>();
        }

        public FileSet(IEnumerable<MGFile> c) {
            m_Items = new List<MGFile>(c);
            m_Items.Sort(new MGFileIDComparer());
        }

        #region IEnumerable<MGFile> Members

        public IEnumerator<MGFile> GetEnumerator() {
            return m_Items.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator() {
            return GetEnumerator1();
        }

        #endregion

        public void Add(MGFile file) {
            m_Items.Add(file);
            m_Items.Sort(new MGFileIDComparer());
        }

        public IEnumerable<MGTag> GetAllTags(string tagName) {
            throw new NotImplementedException();
            //Dim tags As New List(Of MGTag)
            //For Each file As MGFile In m_Items
            //    tags.Add(file.Tags.Item(tagName))
            //Next
            //Return From t In tags Group By t.Value Into tg = First() Select tg
        }

        public IEnumerator GetEnumerator1() {
            return m_Items.GetEnumerator();
        }

        public static FileSet Intesect(FileSet left, FileSet right) {
            var common = new FileSet();

            int l = 0;
            int r = 0;
            while ((l < left.m_Items.Count && r < right.m_Items.Count)) {
                if (left.m_Items[l].ID == right.m_Items[r].ID) {
                    common.m_Items.Add(left.m_Items[l]);
                    l += 1;
                    l += 1;
                } else if (left.m_Items[l].ID.CompareTo(right.m_Items[r].ID) < 0)
                    l += 1;
                else
                    r += 1;
            }

            return common;
        }
    }
}